numba/targets/optional.py

```diff
@@ -89,6 +85,7 @@ def optional_to_optional(context, builder, fromty, toty, val):
 
 
 @lower_cast(types.Any, types.Optional)
+@lower_cast(types.Boolean, types.Optional)
```

Comment:
Good catch.  I didn't think about it.  The ambiguity comes from ``cast(bool, any)``, which I should actually restrict.